include/libcyphal/runnable.hpp

```diff
@@ -0,0 +1,31 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_RUNNABLE_HPP_INCLUDED
+#define LIBCYPHAL_RUNNABLE_HPP_INCLUDED
+
+#include "types.hpp"
+
+namespace libcyphal
+{
+
+class IRunnable
+{
+public:
+    IRunnable(IRunnable&&)                 = delete;
+    IRunnable(const IRunnable&)            = delete;
+    IRunnable& operator=(IRunnable&&)      = delete;
+    IRunnable& operator=(const IRunnable&) = delete;
+
+    virtual void run(const TimePoint now) = 0;
+
+protected:
+    IRunnable()          = default;
+    virtual ~IRunnable() = default;
```

Comment:
It's sorta a nitpick but per [Core Cpp Guidelines C35](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c35-a-base-class-destructor-should-be-either-public-and-virtual-or-protected-and-non-virtual) the `~IRunnable()` shouldn't be virtual. Can we establish this as a standard in this codebase too?

Reply:
It's a bit tricky (for me at least) to decide when we want destructor accessible and when we don't... Even for interfaces we currently do move them into unique_ptr<Concrete> → unique_ptr<Interface> which will require them public.

Reply:
We shouldn't manage the lifecycle of an object through it's pointer. This can lead to dangerous assumptions and amputations (https://godbolt.org/z/1Tj4nzehd). This is what the ~`voldemort_ptr`~, ~`dark-ptr`~, `interface_ptr` type is meant to solve.

Reply:
Thank You Scott for very representable example (with legless dog). Please see latest commit where I moved destructors for "external" entities (`IMedia`-s and `IMultiplexer`) to protected area. BUT, I can't currently do the same for `IRunnable` up to various `IXxxSession`-s we return to user as unique ptrs. This is b/c of our current pmr implementation of UniquePtr - it REQUIRES currently that ~Interface will be available for deleter. The only way I see to remove such requirement is to have type-erased deleter which will capture in a closure original concrete type allocator. It will require `cetl::function` (not yet available). IMHO it has to be type-erased - otherwise we can't have proper `UniquePtr<IX>` → `UniquePtr<IY>` conversions (assuming of course that `X` and `Y` convertible). I discussed this with Pavel - it seems that he agrees with me.

BTW, such type-erased deleter should also solve current standing issue https://github.com/OpenCyphal/CETL/issues/118 (see 2nd bullet)

Reply:
Okay. Let's track this with TODOs in the code. Perhaps I'll take on this work in CETL. It'll give me a reason to move my godbolt example into github.